// @sanity/ui/dist/es/primitives/textInput/textInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { CloseIcon } from '@sanity/icons';
import React, { createElement, forwardRef, isValidElement, useCallback, useMemo } from 'react';
import { isValidElementType } from 'react-is';
import styled from 'styled-components';
import { EMPTY_RECORD } from '../../constants';
import { useForwardedRef, useCustomValidity, useResponsiveProp } from '../../hooks';
import { responsiveRadiusStyle, responsiveInputPaddingStyle, textInputStyle, } from '../../styles/internal';
import { Box } from '../box';
import { Button } from '../button';
import { Card } from '../card';
import { Text } from '../text';
const CLEAR_BUTTON_BOX_STYLE = { zIndex: 2 };
const Root = styled.span(textInputStyle.root);
const InputRoot = styled.span `
  flex: 1;
  min-width: 0;
  display: block;
  position: relative;
`;
const Prefix = styled(Card).attrs({ forwardedAs: 'span' }) `
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;

  & > span {
    display: block;
    margin: -1px;
  }
`;
const Suffix = styled(Card).attrs({ forwardedAs: 'span' }) `
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;

  & > span {
    display: block;
    margin: -1px;
  }
`;
const Input = styled.input(responsiveInputPaddingStyle, textInputStyle.input);
const Presentation = styled.span(responsiveRadiusStyle, textInputStyle.representation);
const LeftBox = styled(Box) `
  position: absolute;
  top: 0;
  left: 0;
`;
const RightBox = styled(Box) `
  position: absolute;
  top: 0;
  right: 0;
`;
export const TextInput = forwardRef((props, forwardedRef) => {
    const { border = true, clearButton, disabled = false, fontSize = 2, icon, iconRight, onClear, padding: paddingProp = 3, prefix, radius = 1, space = 3, suffix, customValidity, type = 'text' } = props, restProps = __rest(props, ["border", "clearButton", "disabled", "fontSize", "icon", "iconRight", "onClear", "padding", "prefix", "radius", "space", "suffix", "customValidity", "type"]);
    const ref = useForwardedRef(forwardedRef);
    const padding = useResponsiveProp(paddingProp);
    // Transient properties
    const $hasClearButton = Boolean(clearButton);
    const $hasIcon = Boolean(icon);
    const $hasIconRight = Boolean(iconRight);
    const $hasSuffix = Boolean(suffix);
    const $hasPrefix = Boolean(prefix);
    useCustomValidity(ref, customValidity);
    // Prevent the clear button from taking the focus away from the input
    const handleClearMouseDown = useCallback((event) => {
        event.preventDefault();
        event.stopPropagation();
    }, []);
    const handleClearClick = useCallback((event) => {
        var _a;
        event.preventDefault();
        event.stopPropagation();
        if (onClear)
            onClear();
        // Focus the input, in case focus has been lost when clicking the clear button
        (_a = ref.current) === null || _a === void 0 ? void 0 : _a.focus();
    }, [onClear, ref]);
    // Render prefix (memoized)
    const prefixNode = useMemo(() => prefix && (React.createElement(Prefix, { borderTop: true, borderLeft: true, borderBottom: true, radius: radius, sizing: "border" },
        React.createElement("span", null, prefix))), [prefix, radius]);
    // Render presentation (memoized)
    const presentationNode = useMemo(() => (React.createElement(Presentation, { "$border": border, "$hasPrefix": $hasPrefix, "$hasSuffix": $hasSuffix, "$radius": radius },
        icon && (React.createElement(LeftBox, { padding: padding },
            React.createElement(Text, { size: fontSize },
                isValidElement(icon) && icon,
                isValidElementType(icon) && createElement(icon)))),
        !$hasClearButton && iconRight && (React.createElement(RightBox, { padding: padding },
            React.createElement(Text, { size: fontSize },
                isValidElement(iconRight) && iconRight,
                isValidElementType(iconRight) && createElement(iconRight)))))), [border, fontSize, icon, iconRight, padding, radius, $hasClearButton, $hasPrefix, $hasSuffix]);
    // Render clear button (memoized)
    const clearButtonBoxPadding = useMemo(() => padding.map((v) => v - 2), [padding]);
    const clearButtonPadding = useMemo(() => padding.map((v) => v - 1), [padding]);
    const clearButtonProps = useMemo(() => (typeof clearButton === 'object' ? clearButton : EMPTY_RECORD), [clearButton]);
    const clearButtonNode = useMemo(() => clearButton && (React.createElement(RightBox, { padding: clearButtonBoxPadding, style: CLEAR_BUTTON_BOX_STYLE },
        React.createElement(Button, Object.assign({}, clearButtonProps, { "data-qa": "clear-button", fontSize: fontSize, icon: CloseIcon, mode: "bleed", onClick: handleClearClick, onMouseDown: handleClearMouseDown, padding: clearButtonPadding })))), [
        clearButton,
        clearButtonBoxPadding,
        clearButtonPadding,
        clearButtonProps,
        fontSize,
        handleClearClick,
        handleClearMouseDown,
    ]);
    // Render suffix (memoized)
    const suffixNode = useMemo(() => suffix && (React.createElement(Suffix, { borderTop: true, borderRight: true, borderBottom: true, radius: radius, sizing: "border" },
        React.createElement("span", null, suffix))), [radius, suffix]);
    return (React.createElement(Root, { "data-ui": "TextInput" },
        prefixNode,
        React.createElement(InputRoot, null,
            React.createElement(Input, Object.assign({ "data-as": "input" }, restProps, { "$iconLeft": $hasIcon, "$iconRight": $hasIconRight || $hasClearButton, "$padding": padding, "$space": space, "$fontSize": fontSize, disabled: disabled, ref: ref, type: type })),
            presentationNode,
            clearButtonNode),
        suffixNode));
});
TextInput.displayName = 'TextInput';
