// @sanity/form-builder/lib/inputs/DateInputs/CommonDateTimeInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommonDateTimeInput = void 0;

var _react = _interopRequireDefault(require("react"));

var _components = require("@sanity/base/components");

var _autoId = require("@reach/auto-id");

var _ui = require("@sanity/ui");

var _DateTimeInput = require("./base/DateTimeInput");

var _legacy = require("./legacy");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var DEFAULT_PLACEHOLDER_TIME = new Date();

var CommonDateTimeInput = /*#__PURE__*/_react.default.forwardRef(function CommonDateTimeInput(props, forwardedRef) {
  var value = props.value,
      markers = props.markers,
      title = props.title,
      description = props.description,
      placeholder = props.placeholder,
      inputFormat = props.inputFormat,
      selectTime = props.selectTime,
      timeStep = props.timeStep,
      readOnly = props.readOnly,
      level = props.level,
      presence = props.presence,
      onChange = props.onChange,
      rest = _objectWithoutProperties(props, ["value", "markers", "title", "description", "placeholder", "inputFormat", "selectTime", "timeStep", "readOnly", "level", "presence", "onChange"]);

  var _React$useState = _react.default.useState(),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      parseError = _React$useState2[0],
      setParseError = _React$useState2[1];

  var _React$useState3 = _react.default.useState(),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      inputValue = _React$useState4[0],
      setInputValue = _React$useState4[1];

  var handleDatePickerInputChange = _react.default.useCallback(event => {
    var nextInputValue = event.currentTarget.value;
    setParseError(undefined);
    setInputValue(undefined);

    if (!nextInputValue) {
      onChange(null);
      return;
    }

    var result = (0, _legacy.parse)(nextInputValue, inputFormat);

    if (result.isValid) {
      onChange(result.date);
    } else {
      setParseError(result.error); // keep the input value floating around as long as it's invalid so that the
      // user can continue to edit

      setInputValue(nextInputValue);
    }
  }, [onChange, inputFormat]);

  var handleDatePickerChange = _react.default.useCallback(nextDate => {
    onChange(nextDate);
    setParseError(undefined);
    setInputValue(undefined);
  }, [onChange]);

  var inputRef = (0, _ui.useForwardedRef)(forwardedRef);
  var id = (0, _autoId.useId)();
  var valueAsDate = value && new Date(value);
  var textInputValue = inputValue ? inputValue : valueAsDate ? (0, _legacy.format)(valueAsDate, inputFormat) : '';
  return /*#__PURE__*/_react.default.createElement(_components.FormField, {
    __unstable_markers: parseError ? [...markers, {
      type: 'validation',
      level: 'error',
      item: {
        message: parseError,
        paths: []
      }
    } // casting to marker to avoid having to implement cloneWithMessage on item
    ] : markers,
    title: title,
    level: level,
    description: description,
    __unstable_presence: presence,
    inputId: id
  }, readOnly ? /*#__PURE__*/_react.default.createElement(_ui.TextInput, {
    value: textInputValue,
    disabled: true
  }) : /*#__PURE__*/_react.default.createElement(_DateTimeInput.DateTimeInput, _extends({}, rest, {
    id: id,
    selectTime: selectTime,
    timeStep: timeStep,
    placeholder: placeholder || "e.g. ".concat((0, _legacy.format)(DEFAULT_PLACEHOLDER_TIME, inputFormat)),
    ref: inputRef,
    value: valueAsDate,
    inputValue: inputValue ? inputValue : valueAsDate ? (0, _legacy.format)(valueAsDate, inputFormat) : '',
    readOnly: readOnly,
    onInputChange: handleDatePickerInputChange,
    onChange: handleDatePickerChange,
    customValidity: parseError
  })));
});

exports.CommonDateTimeInput = CommonDateTimeInput;
