// @sanity/ui/dist/cjs/helpers/element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHTMLTextAreaElement = exports.isHTMLSelectElement = exports.isHTMLButtonElement = exports.isHTMLInputElement = exports.isHTMLAnchorElement = exports.isHTMLElement = void 0;
function isHTMLElement(node) {
    return node instanceof Node && node.nodeType === Node.ELEMENT_NODE;
}
exports.isHTMLElement = isHTMLElement;
function isHTMLAnchorElement(element) {
    return isHTMLElement(element) && element.nodeName === 'A';
}
exports.isHTMLAnchorElement = isHTMLAnchorElement;
function isHTMLInputElement(element) {
    return isHTMLElement(element) && element.nodeName === 'INPUT';
}
exports.isHTMLInputElement = isHTMLInputElement;
function isHTMLButtonElement(element) {
    return isHTMLElement(element) && element.nodeName === 'BUTTON';
}
exports.isHTMLButtonElement = isHTMLButtonElement;
function isHTMLSelectElement(element) {
    return isHTMLElement(element) && element.nodeName === 'SELECT';
}
exports.isHTMLSelectElement = isHTMLSelectElement;
function isHTMLTextAreaElement(element) {
    return isHTMLElement(element) && element.nodeName === 'TEXTAREA';
}
exports.isHTMLTextAreaElement = isHTMLTextAreaElement;
